pkg/monitoring/vmistats/collector.go

```diff
@@ -171,16 +184,156 @@ func checkNonEvictableVMAndSetMetric(vmi *k6tv1.VirtualMachineInstance) float64
 	return setVal
 }
 
-func updateVMIEvictionBlocker(vmis []*k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
+func updateVMIMetrics(vmis []*k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
 	for _, vmi := range vmis {
-		mv, err := prometheus.NewConstMetric(
-			vmiEvictionBlockerDesc, prometheus.GaugeValue,
-			checkNonEvictableVMAndSetMetric(vmi),
-			vmi.Status.NodeName, vmi.Namespace, vmi.Name,
-		)
-		if err != nil {
-			continue
+		updateVMIEvictionBlocker(vmi, ch)
+	}
+}
+
+func updateVMIEvictionBlocker(vmi *k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
+	mv, err := prometheus.NewConstMetric(
+		vmiEvictionBlockerDesc, prometheus.GaugeValue,
+		checkNonEvictableVMAndSetMetric(vmi),
+		vmi.Status.NodeName, vmi.Namespace, vmi.Name,
+	)
+	if err != nil {
+		return
+	}
+	ch <- mv
+
+}
+
+func getTransitionTimeSeconds(fromCreation bool, oldVMI *k6tv1.VirtualMachineInstance, newVMI *k6tv1.VirtualMachineInstance) (float64, error) {
+
+	var oldTime *metav1.Time
+	var newTime *metav1.Time
+
+	if fromCreation || oldVMI == nil || (oldVMI.Status.Phase == k6tv1.VmPhaseUnset) {
+		oldTime = newVMI.CreationTimestamp.DeepCopy()
+	}
+
+	for _, transitionTimestamp := range newVMI.Status.PhaseTransitionTimestamps {
+		if newTime == nil && transitionTimestamp.Phase == newVMI.Status.Phase {
+			newTime = transitionTimestamp.PhaseTransitionTimestamp.DeepCopy()
+		} else if oldTime == nil && oldVMI != nil && transitionTimestamp.Phase == oldVMI.Status.Phase {
+			oldTime = transitionTimestamp.PhaseTransitionTimestamp.DeepCopy()
+		} else if oldTime != nil && newTime != nil {
+			break
 		}
-		ch <- mv
 	}
+
+	if newTime == nil || oldTime == nil {
+		// no phase transition timestamp found
+		return 0.0, fmt.Errorf("missing phase transition timestamp")
+	}
+
+	diffSeconds := newTime.Time.Sub(oldTime.Time).Seconds()
+
+	// when transitions are very fast, we can encounter time skew. Make 0 the floor
+	if diffSeconds < 0 {
+		diffSeconds = 0.0
+	}
+
+	return diffSeconds, nil
+}
+
+func phaseTransitionTimeBuckets() []float64 {
+	return []float64{
+		(1 * time.Second.Seconds()),
```

Comment:
I would suggest to also have a millisecond precision in the histogram.
There are 12 buckets now, but we could make it 15...
`(0.1 * time.Second.Seconds()),`
`(0.5 * time.Second.Seconds()),`
Or just the 0.5....

Reply:
updated, i added a .5 and then some more fine granular buckets in the middle. < .5 isn't that useful to us i don't think.